java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -446,29 +397,26 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
     // look in site members
     for (JavaSymbol symbol : site.getSymbol().members().lookup(name)) {
       if (symbol.kind == JavaSymbol.MTH) {
-        JavaSymbol best = selectBest(env, callSite, argTypes, symbol, bestSoFar.symbol, autoboxing);
-        if(best == symbol) {
-          bestSoFar = Resolution.resolution(best);
-          bestSoFar.type = resolveTypeSubstitution(((JavaType.MethodJavaType) best.type).resultType, site);
-          JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) best;
-          bestSoFar.type = handleTypeArguments(typeParams, bestSoFar.type, methodSymbol);
+        Resolution best = selectBest(env, callSite, argTypes, typeParams, symbol, bestSoFar, autoboxing);
+        if (best.symbol == symbol) {
+          bestSoFar = best;
         }
       }
     }
     //look in supertypes for more specialized method (overloading).
     if (superclass != null) {
       Resolution method = findMethod(env, callSite, superclass, name, argTypes, typeParams);
-      method.type = resolveTypeSubstitution(resolveTypeSubstitution(method.type, superclass), site);
-      JavaSymbol best = selectBest(env, callSite, argTypes, method.symbol, bestSoFar.symbol, autoboxing);
-      if(best == method.symbol) {
+      method.type = typeSubstitutionSolver.applySiteSubstitution(method.type, site, superclass);
+      Resolution best = selectBest(env, callSite, argTypes, typeParams, method.symbol, bestSoFar, autoboxing);
+      if (best.symbol == method.symbol) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)